controllers/pruner/pruner_controller.go

```diff
@@ -0,0 +1,361 @@
+
+	"github.com/operator-framework/operator-lib/prune"
+	"github.com/robfig/cron/v3"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// DevWorkspacePrunerReconciler reconciles DevWorkspace objects for pruning purposes.
+type DevWorkspacePrunerReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+
+	cron *cron.Cron
+}
+
+func shouldReconcileOnUpdate(e event.UpdateEvent, log logr.Logger) bool {
+	log.Info("DevWorkspaceOperatorConfig update event received")
+	oldConfig, ok := e.ObjectOld.(*controllerv1alpha1.DevWorkspaceOperatorConfig)
+	if !ok {
+		return false
+	}
+	newConfig, ok := e.ObjectNew.(*controllerv1alpha1.DevWorkspaceOperatorConfig)
+	if !ok {
+		return false
+	}
+
+	oldCleanup := oldConfig.Config.Workspace.CleanupCronJob
+	newCleanup := newConfig.Config.Workspace.CleanupCronJob
+
+	differentBool := func(a, b *bool) bool {
+		switch {
+		case a == nil && b == nil:
+			return false
+		case a == nil || b == nil:
+			return true
+		default:
+			return *a != *b
+		}
+	}
+	differentInt32 := func(a, b *int32) bool {
+		switch {
+		case a == nil && b == nil:
+			return false
+		case a == nil || b == nil:
+			return true
+		default:
+			return *a != *b
+		}
+	}
+
+	if oldCleanup == nil && newCleanup == nil {
+		return false
+	}
+	if (oldCleanup == nil && newCleanup != nil) || (oldCleanup != nil && newCleanup == nil) {
+		return true
+	}
+	if differentBool(oldCleanup.Enable, newCleanup.Enable) {
+		return true
+	}
+	if differentBool(oldCleanup.DryRun, newCleanup.DryRun) {
+		return true
+	}
+	if differentInt32(oldCleanup.RetainTime, newCleanup.RetainTime) {
+		return true
+	}
+	return oldCleanup.Schedule != newCleanup.Schedule
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *DevWorkspacePrunerReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	log := r.Log.WithName("setupWithManager")
+	log.Info("Setting up DevWorkspacePrunerReconciler")
+
+	configPredicate := predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return shouldReconcileOnUpdate(e, log)
+		},
+		CreateFunc:  func(e event.CreateEvent) bool { return true },
+		DeleteFunc:  func(e event.DeleteEvent) bool { return true },
+		GenericFunc: func(e event.GenericEvent) bool { return false },
+	}
+
+	maxConcurrentReconciles, err := config.GetMaxConcurrentReconciles()
+	if err != nil {
+		return err
+	}
+
+	// Initialize cron scheduler
+	r.cron = cron.New()
+
+	return ctrl.NewControllerManagedBy(mgr).
+		WithOptions(controller.Options{MaxConcurrentReconciles: maxConcurrentReconciles}).
+		Named("DevWorkspacePruner").
+		Watches(&source.Kind{Type: &controllerv1alpha1.DevWorkspaceOperatorConfig{}},
+			handler.EnqueueRequestsFromMapFunc(func(object client.Object) []ctrl.Request {
+				objectNamespace := object.GetNamespace()
+				operatorNamespace, err := infrastructure.GetNamespace()
+
+				// Ignore events from other namespaces
+				if err != nil || objectNamespace != operatorNamespace {
```

Comment:
Could we also check name of the object to make sure it matches https://github.com/akurinnoy/devworkspace-operator/blob/f977dfbeea80f677aae7d997182234b83c41b7c0/pkg/config/sync.go#L44 ?